### PR TITLE
Change MAP and ROW stack types

### PR DESCRIPTION
**Pull request by dain**
## Description

This PR changes `MAP` and `ROW` stack types to be `SqlMap` and `SqlRow` respectively.  The new types do not extend block, and simply wrappers around the underlying column blocks.  Both of these value classes have accessors for the underlying blocks and the offset into the blocks.  

This PR contains a number of cleanup commit to simplify this conversion, and it also contains some minor fixes I found while making this change.

This is part of  [Project Hummingbird](https://github.com/trinodb/trino/issues/14237)

## Release notes

( ) Release notes are required, with the following suggested text:

```markdown
# Section
* The Java stack type for a `MAP` value is now `SqlMap`, which does not implement `Block`. ({issue}`18948`)
* The Java stack type for a `ROW` value is now `SqlRow`, which does not implement `Block`. ({issue}`18948`)
```

### CODE REVIEW

File: core/trino-main/src/main/java/io/trino/operator/scalar/MapConstructor.java [hunk 1]
```diff
@@ -98,16 +99,17 @@ public SpecializedSqlScalarFunction specialize(BoundSignature boundSignature, Fu
                 simpleConvention(FAIL_ON_NULL, NEVER_NULL)).getMethodHandle();
         MethodHandle instanceFactory = constructorMethodHandle(State.class, MapType.class).bindTo(mapType);
 
+        MethodHandle methodHandle = METHOD_HANDLE.bindTo(mapType).bindTo(keyIndeterminate);
         return new ChoicesSpecializedSqlScalarFunction(
                 boundSignature,
                 FAIL_ON_NULL,
                 ImmutableList.of(NEVER_NULL, NEVER_NULL),
-                METHOD_HANDLE.bindTo(mapType).bindTo(keyIndeterminate),
+                methodHandle,
```
Comment (electrum) [inline core/trino-main/src/main/java/io/trino/operator/scalar/MapConstructor.java:108]: This code is the same as the original

File: core/trino-main/src/test/java/io/trino/util/StructuralTestUtil.java [hunk 2]
```diff
@@ -52,7 +53,7 @@ public static Block arrayBlockOf(Type elementType, Object... values)
         return blockBuilder.build();
     }
 
-    public static Block mapBlockOf(Type keyType, Type valueType, Map<?, ?> map)
+    public static SqlMap sqlMapOf(Type keyType, Type valueType, Map<?, ?> map)
```
Comment (electrum) [inline core/trino-main/src/test/java/io/trino/util/StructuralTestUtil.java:58]: You could rename this method in a precursor commit, to make the diff smaller

**Review by electrum — APPROVED** [review]

File: core/trino-spi/src/main/java/io/trino/spi/block/SqlMap.java [hunk 3]
```diff
@@ -47,16 +50,33 @@ public class SqlMap
...
-    public SqlMap(MapType mapType, Block rawKeyBlock, Block rawValueBlock, HashTableSupplier hashTablesSupplier, int offset, int positionCount)
+    public SqlMap(MapType mapType, HashBuildMode mode, Block keyBlock, Block valueBlock)
     {
-        this.mapType = mapType;
-        this.rawKeyBlock = rawKeyBlock;
-        this.rawValueBlock = rawValueBlock;
-        this.hashTablesSupplier = hashTablesSupplier;
+        this.mapType = requireNonNull(mapType, "mapType is null");
+        if (keyBlock.getPositionCount() != valueBlock.getPositionCount()) {
+            throw new TrinoException(GENERIC_INTERNAL_ERROR, format("Key and value blocks have different size: %s %s", keyBlock.getPositionCount(), valueBlock.getPositionCount()));
```
[the remark this answers is not in the document]
Comment (dain) [reply core/trino-spi/src/main/java/io/trino/spi/block/SqlMap.java:53]: The SPI doesn't have verify, but I'll use illegal argument.

File: core/trino-spi/pom.xml [hunk 4]
```diff
@@ -426,6 +426,21 @@
                                     <code>java.class.removed</code>
                                     <old>class io.trino.spi.block.SingleRowBlockEncoding</old>
                                 </item>
+                                <item>
```
[the remark this answers is not in the document]
Comment (dain) [reply core/trino-spi/pom.xml:429]: I believe so.  The CI builds every commit individually.

**Pull request merged**